File: amethyst.ui/src/types/Ruleset.ts
export type RulesState = {
    rules: Ruleset;
}

export type Ruleset = {
    periodRules: PeriodRules;
    jamRules: JamRules;
    lineupRules: LineupRules;
    timeoutRules: TimeoutRules;
    penaltyRules: PenaltyRules;
    intermissionRules: IntermissionRules;
}

export type PeriodRules = {
    periodCount: number;
    durationInSeconds: number;
    periodEndBehavior: PeriodEndBehavior;
}

export type JamRules = {
    resetJamNumbersBetweenPeriods: boolean;
    durationInSeconds: number;
}

export type LineupRules = {
    durationInSeconds: number;
    overtimeDurationInSeconds: number;
}

export type TimeoutRules = {
    teamTimeoutDurationInSeconds: number;
    periodClockBehavior: TimeoutPeriodClockStopBehavior;
    teamTimeoutAllowance: number;
    resetBehavior: TimeoutResetBehavior;
}

export type PenaltyRules = {
    foulOutPenaltyCount: number;
}

export type IntermissionRules = {
    durationInSeconds: number;
}

export enum PeriodEndBehavior {
    Immediately = "Immediately",
    AnytimeOutsideJam = "AnytimeOutsideJam",
    OnJamEnd = "OnJamEnd",
    Manual = "Manual",
}

export enum TimeoutPeriodClockStopBehavior {
    OfficialTimeout = 1 << 0,
    TeamTimeout = 1 << 1,
    OfficialReview = 1 << 2,
    All = OfficialTimeout | TeamTimeout | OfficialReview,
}

export enum TimeoutResetBehavior {
    Never = "Never",
    Period = "Period",
}

File: jamster.engine/Domain/Ruleset.cs
using System.Text.Json.Serialization;

namespace jamster.Domain;

public record Ruleset(
    PeriodRules PeriodRules,
    JamRules JamRules,
    LineupRules LineupRules,
    TimeoutRules TimeoutRules,
    PenaltyRules PenaltyRules,
    IntermissionRules IntermissionRules,
    InjuryRules InjuryRules
);

public record PeriodRules(
    int PeriodCount,
    int DurationInSeconds,
    PeriodEndBehavior PeriodEndBehavior
);

public record JamRules(
    bool ResetJamNumbersBetweenPeriods,
    int DurationInSeconds
);

public record LineupRules(
    int DurationInSeconds,
    int OvertimeDurationInSeconds
);

public record TimeoutRules(
    int TeamTimeoutDurationInSeconds,
    TimeoutPeriodClockStopBehavior PeriodClockBehavior,
    int TeamTimeoutAllowance,
    TimeoutResetBehavior ResetBehavior
);

public record PenaltyRules(
    int FoulOutPenaltyCount
);

public record IntermissionRules(
    int DurationInSeconds
);

public record InjuryRules(
    int JamsToSitOutFollowingInjury,
    int NumberOfAllowableInjuriesPerPeriod
);

public enum PeriodEndBehavior
{
    Immediately,
    AnytimeOutsideJam,
    OnJamEnd,
    Manual,
}

[Flags]
public enum TimeoutPeriodClockStopBehavior
{
    All = OfficialTimeout | TeamTimeout | OfficialReview,
    OfficialTimeout = 1 << 0,
    TeamTimeout = 1 << 1,
    OfficialReview = 1 << 2,
}

public enum TimeoutResetBehavior
{
    Never,
    Period,
}